Autofire timeout

Aiden166 opened the pull request:
More of a request to test than to merge. I am currently unable to do any tests myself. I was wondering if someone could look over this. It *should* work but I'm not confident in it. What it's supposed to do is add the ability for autofire devices to automatically disable themselves for X amount of seconds if they are triggered Y times in Z seconds. X, Y, and Z are options in the config file. There's also an option to disable the timeout entirely. Basically, when disabled it shouldn't change anything. Sorry for anything I did wrong. I'm new to GitHub and okay-ish at Python. Thanks in advance.

Aiden166 committed:
Update config_spec.py

Added config options for enabling/disabling autofire self timeouts (enable_timeouts), setting the time period to watch for (timeout_watch_time), setting the maximum number of hits allowed in that time period (timeout_max_hits), and the time to disable the device for if the maximum amount of hits are exceeded in that time period (timeout_disable_time).

Aiden166 committed:
Added ability to have autofires self timeout.

It should now be possible to have autofire devices automatically disable themselves for a configurable amount of seconds if they are activated a configurable amount of times in a configurable amount of seconds. Basically if activated X times in Y seconds, disable for Z seconds. X, Y, and Z are options in the config file.

Aiden166 committed:
Changed types.

Changed timeout_watch_time and timeout_disable_time in autofire_coils to type 'ms' from invalid type 's'.

Aiden166 committed:
Accommodated unit of time.

Changed _timeout_watch_time and _timeout_disable_time definitions to divide by 1000 do accommodate ms input instead of s input.

Aiden166 commented:
Let's hope that fixes it.

Aiden166 commented:
I guess not.

jabdoa2 commented on mpf/devices/autofire.py:
```diff
@@ -47,6 +47,13 @@ def _initialize(self) -> None:
+            self._timeout_max_hits = self.config['timeout_max_hits']
+            self._timeout_disable_time = self.config['timeout_disable_time']/1000
+            self._timeout_hits = []
+            from time import time as _time
```
Can you move the import to the top of the file? Pylint complains about this.

jabdoa2 commented on mpf/devices/autofire.py:
```diff
@@ -47,6 +47,13 @@ def _initialize(self) -> None:
+            self._timeout_disable_time = self.config['timeout_disable_time']/1000
+            self._timeout_hits = []
+            from time import time as _time
+            from threading import Timer as _Timer
```
This timer will interfere with asyncio. You can use self.machine.clock instead or instantiate DelayManager(self.machine.delayRegistry) for a nicer interface.

jabdoa2 commented on mpf/devices/autofire.py:
```diff
@@ -110,6 +117,17 @@ def _hit(self):
+                    break
+            if len(self._timeout_hits)>self._timeout_max_hits:
+                self.disable()
+                _timeout_timer=_Timer(self._timeout_disable_time,self.enable)
```
I like the approach. There is a race if the pops are disabled during the disable period. With DelayManager you could just cancel the delay in the disable function.

jabdoa2 reviewed (dismissed):
Thanks for your efforts! The general approach is good. A test would be nice. You can just run:

pip3 install -e .

Within the mpf directory to get everything installed. Then you can run unit tests via:

python3 -m unittest discover -s mpf.tests

See: http://docs.missionpinball.org/en/dev/about/contributing_to_mpf.html (you can ignore all the mc related points if you only make changes to mpf).

jabdoa2 reviewed (commented):
Your test could go here: https://github.com/missionpinball/mpf/blob/dev/mpf/tests/test_Autofire.py

Aiden166 committed:
Moved import, changed Timer to delay, and fixed race.

Moved "from time import time as _time" from inside the _initialize method to the top of the file so Pylint is happy. Removed Timer from threading and added delay from DelayManger in its place so it doesn't interfere with asyncio. Added delay removal to disable method so as to prevent race.

Aiden166 commented:
I fixed the import, changed the Timer to a delay, and fixed the race condition. I think. About the tests, I am currently working on a desktop at my school, which means I have to do everything in-browser as I have limited conditions. I have Python installed but it's 3.6 and the installation help says it will only work with 3.4. Additionally, I can't access the command line at all because it's been disabled. I am unable to do any tests at this time. I should be able to once I get home, but I need to find a way to be able to call Python 3.4 from the command line instead of 3.6. I've looked at the automated tests, but it seems to be unable to find the config options I added, so I need to figure out what I screwed up on that, which is what is giving me the most trouble right now. I will try to write some tests for it but I'll need to figure out how first. Thanks for all the help you've given me, and sorry to trouble you so much.

Aiden166 committed:
Added enable_timeout

Added enable_timeout to see if that fixes the system's inability to find it in the config.

Aiden166 committed:
Removed enable_timeout

Removed enable_timeout because that only made it worse.

jabdoa2 commented:
Python 3.6 should be fine for MPF. I think it currently is only an issue with Kivy on Windows.

Linter is currently unhappy on travis:

```
mpf/devices/autofire.py
  Line: 9
    pylint: wrong-import-order / standard import "from time import time as _time" should be placed before "from mpf.core.delays import DelayManager"
  Line: 53
    pep8: E226 / missing whitespace around arithmetic operator (col 73)
    pylint: attribute-defined-outside-init / Attribute '_timeout_watch_time' defined outside __init__ (col 12)
  Line: 54
    pylint: attribute-defined-outside-init / Attribute '_timeout_max_hits' defined outside __init__ (col 12)
  Line: 55
    pylint: attribute-defined-outside-init / Attribute '_timeout_disable_time' defined outside __init__ (col 12)
  Line: 56
    pylint: attribute-defined-outside-init / Attribute '_timeout_hits' defined outside __init__ (col 12)
  Line: 108
    pep8: W293 / blank line contains whitespace (col 1)
  Line: 125
    pep8: E226 / missing whitespace around arithmetic operator (col 42)
    pep8: E225 / missing whitespace around operator (col 64)
  Line: 129
    pep8: E225 / missing whitespace around operator (col 39)
  Line: 131
    pep8: E231 / missing whitespace after ',' (col 58)
```

Aiden166 committed:
Moved things and added spaces for style

Moved "from time import time as _time" to above "from mpf.core.delays import DelayManager".
Moved "if self.config['enable_timeouts']: ... self._timeout_hits = []" from _initialize method to __init__ method.
Added spaces in all the places I forgot to put spaces. (Sometimes I forget style guides exist)

Aiden166 commented:
Fixed the style. Still have no idea why it can't find "enable_timeouts" in the config file.

Aiden166 committed:
Added enable_timeouts

Tried adding enable_timeouts again because last time I added 'enable_timeout' instead of 'enable_timeouts'. Forgot the 's'.

Aiden166 commented:
Okay, I have no idea why it can't find it. If someone could tell me why it's unable to find the option, I would be extremely grateful.

Aiden166 committed:
Removed "enable_timeouts"

Removed "enable_timeouts" because it didn't help. At least it didn't make it worse this time.

jabdoa2 commented on mpf/devices/autofire.py:
```diff
@@ -40,6 +42,11 @@ def __init__(self, machine: "MachineController", name: str) -> None:
         super().__init__(machine, name)
         self.delay = DelayManager(self.machine.delayRegistry)
         self._ball_search_in_progress = False
+        if self.config['enable_timeouts']:
```
The problem here is that config is set after __init__ is done. I guess all you have to do is move this to _initialize which is called after self.config is set.

Aiden166 replied:
It actually had been in _initialize before and I was still getting the same error. I actually moved it to __init__ because linter was complaining about the variable declarations being outside of __init__ in that error log you posted. I'll try moving it back into _initialize but unfortunately I don't think that's it. It's probably something stupid like I put the options in the wrong config file.

Aiden166 committed:
Moved timeout back to _initialize method.

Moved "if self.config['enable_timeouts']: ... self._timeout_hits = []" back to _initialize method from __init__ method to see if that helps it find the config options.

jabdoa2 committed:
small change to make linter and tests happy

Aiden166 commented:
So, why does that work now? I see that you removed the "enable_timeouts" config check and used the "timeout_watch_time" to decide whether to enable auto-timeout instead, but would you mind explaining to me why that works?

jabdoa2 commented:
That is unrelated. 1. I copied the config to kickbacks which subclass autofires. That fixes the issue with missing config. 2. I added the variable to init to make the linter happy.

Aiden166 commented:
I see. Thank you very much. So I needed to put the config options in "kickbacks" instead of "autofires". Thank you for explaining that to me. Sorry for being so much of a noob.

jabdoa2 committed:
remove now unused variables

jabdoa2 committed:
add annotation

jabdoa2 committed:
fix two logic bugs

jabdoa2 committed:
test timeouts in autofire coils

jabdoa2 commented:
added a test and fixed some issues with the timeout. this should be enough to fix #777

Aiden166 commented:
Should I delete the branch then now?

jabdoa2 commented:
The branch is in your repository so it's up to you. However, there is little reason to keep it.
